### DISCUSSION TITLE
Using Nebula in a manage package?

### DISCUSSION THREAD

**Discussion by dcuenot** [jongpie/NebulaLogger#270] [Q&A] [answered]
Hi @jongpie,

I currently developing a managed package, and I have a stupid question, can I use Nebula to log data from my apex class?

If it's possible, I'm wondering how should I write my APEX code to run & compile in case Nebula is not deployed on my org.

Hoping my questions are clear enough, because I'm still a beginner in the managed package world :)

Thanks a lot for your answer.
Regards
Damien

**Comment by jamessimone** [comment, marked as answer]
@dcuenot hi there! I develop unmanaged packages, but the general strategy I employ you may find helpful - I have a suite of plugins for my package, [including one for Nebula Logger](https://github.com/jamessimone/apex-rollup/tree/main/plugins/NebulaLogger). If you combine that with my [sfdx-project.json](https://github.com/jamessimone/apex-rollup/blob/main/sfdx-project.json), you can see how the Nebula Logger plugin declares a dependency for both packages. By creating a second package, you break the tight coupling that would otherwise occur by declaring Nebula Logger as a dependency for your own package, and give subscribers easy ways to opt into using the functionality surrounding Nebula.

Hopefully that answer helps. I'm also planning (at some point) to write an article & wiki section here about this.

**Reply by jongpie** [reply]
Thanks, @jamessimone! 

@dcuenot, I highly recommend checking out how @jamessimone has handled his optional integration with Nebula Logger - it's a fantastic, real example of how it can be done. One detail I'll add - I believe that his approach will only work with [2nd Generation Packages](https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_dev2gp.htm), which are much better overall than [1st Generation Packages](https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_build_release.htm) anyway, so just make sure you're planning to create a 2nd generation package.

**Reply by dcuenot** [reply]
Thanks @jamessimone & @jongpie, I think I understood how it's working, and I will use this plugin pattern for my use case.

Just one question to be sure I fully understood, as I'm developing a 2GP managed package with a namespace, I'm force to use the managed version of Nebula, I cannot use the unlocked version, right?

**Reply by jongpie** [reply]
@dcuenot that's a great question - I'm not 100% sure, but if it's possible to use the unlocked version, I would highly recommend that as it is more powerful than the managed version. If I have a chance in the new few weeks, I will try to create a demo managed package to see if it's possible to use the unlocked version - if you're able to test this yourself, let me know what you find!

**Reply by dcuenot-sfdc** [reply]
@jongpie On my scratch org, I tried to deployed the unlocked package, but I faced this issue:
`ERROR running force:package:install:  You cannot install an unlocked package without a namespace into an org with a namespace.`

So I'm guessing, it's not possible to use the unlocked package when we are building a manage package (so with a namespace), but as I'm still new in this package world, maybe I did something wrong somewhere else :)

**Reply by jongpie** [reply]
@dcuenot thanks for the insight! You may still be able to use the unlocked version of Nebula Logger, with 2 changes to your process:

- Your "plugin package" may need to be an unlocked package (without a namespace). Your core package can still be a managed package (with a namespace), but the plugin might need to be an unlocked package instead.
- In order to make your plugin package an unlocked package (without a namespace), you'll have to do any development for the plugin package in another scratch org that does _not_ have a namespace. It will also need to have both your managed package, and the unlocked version of Nebula Logger, installed in the scratch org.

I'm not positive that this will work, but I think it's worth trying to see if this will let you use the unlocked version of Nebula Logger. In case it's helpful, here is a (very basic) diagram of how I think it could work.

![image](https://user-images.githubusercontent.com/1267157/151239132-7c6e670a-6620-4553-9fe8-9cf3a27b0b1d.png)

Let me know if you're able to test out this theory!

**Reply by dcuenot** [reply]
I will try to find some time this week-end to explore this way. Will keep you posted
